File: matlab/trainQ.m
function [theta sigma rho J] = trainQ(F, X, L, y, rho, sigma, m, n, l, theta, iter1, alpha_theta, alpha_sigma, alpha_rho, MINnorm, MAXnorm, name, J_thresh)
    [F Sigma] = ImpactQ(X, L, rho, sigma, m, n, l);
%   size(F) = [m L]
%    disp(F);

    fprintf('\n The Q classifier is training... \n');
    % Add intercept term
    F = [ones(m, 1) F];    
    J = zeros(iter1, 1);
    grad_sigma = zeros(n, 1);
    grad_rho = zeros(n, n);
    STOP = 0;% flag for stopping the iterations
    for inc = 1:1:iter1,
      if(STOP==0),
          
          % after each iteration the gradient factor is set to zero
          grad_sigma = zeros(n, 1);
          grad_rho = zeros(n, n);

          % DO THE DIMESION checking again of the grad factors. ....!!! IMP !!!
          if(inc == 1),
            fprintf('\nThe Cost/penalty is printed below\n');
            tic;
            start = tic;
          end
          if(inc == 2),
            time_each_iter = finish;
            fprintf('\nTime remaining for simulation to end\n');
            sectohms( time_each_iter * (iter1 - inc));
          end

          [F(:, 2:end) Sigma tempinv_Sigma] = ImpactQ(X, L, rho, sigma, m, n, l);
          predictions = my_sigmoid(F*theta);
          % size of predictions = [m x 1(one)] 
%          disp(rho);
%          disp(sigma);
%          disp(Sigma);
          fprintf('.');

          if (inc == 1),
            % the cost function or the penalty incurred 
            for i = 1:1:m,
              J(inc) = J(inc) + -1/m * (y(i)*log(predictions(i))+(1-y(i)) * log(1-predictions(i)));  
            end
            J_old = J(inc);
            J_new = J(inc);
            % displaying the cost function 
            fprintf('Cost function = %f\n', J(inc));
          end 

          if(mod(inc, 5)== 0),
            %            fprintf('|');
            if(mod(inc, 20)==0),
                fprintf('\nTime remaining for simulation to end\n');
                sectohms( time_each_iter * (iter1 - inc));
                J_old = J_new;
                % the cost function or the penalty incurred 
                for k = 1:1:m,
                  J(inc) = J(inc) + -1/m * (y(k)*log(predictions(k))+(1-y(k)) * log(1-predictions(k)));
                end
                J_new = J(inc);

                % displaying the cost function 
                fprintf('Cost function = %f\n', J(inc));
                % saving the variables after some set of iterations 
                save(strcat('temp_', name), 'MINnorm', 'MAXnorm','rho', 'sigma', 'theta', 'L', 'J');
                if ((J_old-J_new)<J_thresh),
                  fprintf('Simulation ran for %d number of iterations and difference in J_old - J_new = %f-%f = %f\n', inc, J_old, J_new, J_old-J_new);
                  STOP = 1;
                end
            end
          end
          % ************************************************** gradient descent of sigma **************************************************
%          tempinv_Sigma = zeros(n, n);
%          tempinv_Sigma = pinv(Sigma);
          if (alpha_sigma ~= 0), 
            D = zeros(n, n); % the derivative of Sigma inverse
            % gradient descent for Sigma/sigma
            for j = 1:1:n,% for each feature calculate the gradient of sigma
              % calculating the derivative of Sigma wrt sigma_j
              der_Sigma = zeros(n, n);
              for p = 1:1:n,
                for q = 1:1:n,
                  if p==j && q==j,
                    der_Sigma(p, q) = 2*sigma(j);
                  elseif p==j,
                    der_Sigma(p, q) = rho(p, q)*sigma(q);
                  elseif q==j,
                    der_Sigma(p, q) = rho(p, q)*sigma(p);
                  else,
                    der_Sigma(p, q) = 0; 
                  end 
                end
              end
              D = -1*tempinv_Sigma * der_Sigma * tempinv_Sigma;
  %            D = -1*inv(Sigma) * der_Sigma * inv(Sigma);
              for i = 1:1:m, % iterate over each example
                second_matrix = zeros(l, 1);
                for k = 1:1:l,
                  second_matrix(k) = (X(i, :) - L(k, :))*D*(X(i, :)-L(k, :))'; 
                end
                grad_sigma(j) = grad_sigma(j) + ( y(i)- predictions(i) ) .*F(i, 2:end).*theta(2:end, 1)' * second_matrix ;
              end
            end
          
            % updating the sigma 
            for k = 1:1:n,
              sigma(k) = sigma(k) - alpha_sigma * 1/(2*m) * grad_sigma(k);
            end
          end
          % end: gradient descent for sigma
        
          %********************************************************** gradient descent for rho **************************************************
          if(alpha_rho ~= 0), 
            D_rho = zeros(n, n); % the derivative of Sigma inverse
            % gradient descent for rho
            for j_p = 1:1:n,% for each landmark points calculate the gradient of rho
              for j_q = j_p+1:1:n,
                % calculating the derivative of Sigma wrt rho_p,q
                der_Sigma_rho = zeros(n, n);
                for p = 1:1:n,
                  for q = p+1:1:n,
                    if p==j_p && q==j_q,
                      der_Sigma_rho(p, q) = sigma(p)*sigma(q);
                      der_Sigma_rho(q, p) = der_Sigma_rho(p, q);
                    else,
                      der_Sigma_rho(p, q) = 0;
                      der_Sigma_rho(q, p) = der_Sigma_rho(p, q);
                    end 
                  end
                end
                D_rho = -1 *tempinv_Sigma * der_Sigma_rho * tempinv_Sigma;
  %              D_rho = -1 * inv(Sigma) * der_Sigma_rho * inv(Sigma);
                for i = 1:1:m, % iterate over each example
                  second_matrix = zeros(l, 1);
                  for k = 1:1:l,
                    second_matrix(k) =(X(i, :) - L(k, :))*D_rho*(X(i, :)-L(k, :))'; 
                  end 
                  grad_rho(j_p, j_q) = grad_rho(j_p, j_q) + ( y(i)- predictions(i) ) .*F(i, 2:end).*theta(2:end, 1)' * second_matrix ;
                end
              end
            end
            % updating the rho parameters
            for p = 1:1:n, 
              for q = p+1:1:n, 
                if (p==q),
                  rho(p, q) = 1;
                else,
                  rho(p, q) = rho(p, q) - alpha_rho * (1/(2*m))* grad_rho(p,q);
                  rho(q, p) = rho(p, q);
                end
              end
            end
          end
          % end: gradient descent for rho
          % *********************************** gradient descent of theta ************************************************        
          grad_theta = zeros(l+1, 1);
          for k = 1:1:l+1,
            for i = 1:1:m,
              grad_theta(k) = grad_theta(k) + (-1/m)*(y(i)- predictions(i))* F(i, k); 
            end
          end
          theta = theta - alpha_theta * grad_theta;
          % end: gradient descent for theta
          if (inc == 1),
            finish = toc;
          end
      end
    end
end

File: python/trainQ.py
# This function trains the Q-Classifier using the modified gradient descent. Returns the trained parameters and the cost function
import time, datetime, pickle, sys
import numpy as np
import ImpactQ
def my_sigmoid(x):
    return 1.0/ (1.0 + np.exp(-x))

def train(F, X, L, y, rho, sigma, m, n, l, theta, iter1, alpha_theta, alpha_sigma, alpha_rho, MINnorm, MAXnorm, fileBaseName, J_thresh):
    print 'Training the Q-Classifier...'
    F, Sigma, temp_inv_Sigma = ImpactQ.Impact(X, L, rho, sigma, m, n, l)
    F = np.hstack((np.ones((m, 1), float), F)) # adding a column of 1's to F
    J = np.zeros((iter1, 1), float)
    STOP = 0 # flag to stop simulations 
    for inc in range(0, iter1):
        if (STOP==0):
            grad_sigma = np.zeros((n, 1), float)
            grad_rho = np.zeros((n, n), float)
            if (inc == 0):
                print ' The penalty incurred(Cost function) is shown below'
                start = time.time()
            if (inc == 1):
                time_each_iter = finish-start
                print 'Time remaining for simulation to end'
                print str(datetime.timedelta(seconds = time_each_iter * (iter1-inc)))
            
            F[:, 1::], Sigma, temp_inv_Sigma = ImpactQ.Impact(X, L, rho, sigma, m, n, l)
#            temp_inv_Sigma = np.linalg.pinv(np.matrix(Sigma))
            # setting up a loop to calculate the predictions for each Input points
            predictions = np.zeros((m, 1), float)
            for i in range(0, m):
                predictions[i] = my_sigmoid(F[i, :] * np.matrix(theta).transpose())
            
    #        print(".|."),
            sys.stdout.write('.')
            
            if (inc == 0):
                for i in range(0, m):
                    J[inc] = J[inc] + -1.0/m * (y[i]*np.log(predictions[i])+ (1-y[i]) * np.log(1-predictions[i]))
                J_old = J[inc]
                J_new = J[inc]
                print("Cost function = %f" %(J[inc]))

            if inc%5 == 0 and inc != 0 :
                if inc%20 == 0 :
                    print 'Time remaining for the simulation to end'
                    print str(datetime.timedelta(seconds = time_each_iter * (iter1-inc)))
                    J_old = J_new
                    for i in range(0, m):
                        J[inc] = J[inc] + -1.0/m * (y[i]*np.log(predictions[i])+ (1-y[i]) * np.log(1-predictions[i]))
                    J_new = J[inc]
                    print("Cost function = %f" %(J[inc]))
                    # saving the parameters after certain amount of iterations.
                    with open(fileBaseName+'_param_temp.pickle', 'w') as ft: # temp variables file
                        pickle.dump([MINnorm, MAXnorm, rho, sigma, theta, L, J], ft)
                    if((J_old-J_new)<J_thresh):
                        print('Simulation ran for %d number of iterations and difference in J_old - J_new = %f-%f = %f\n'%( inc, J_old, J_new, J_old-J_new))
                        STOP = 1
            # Gradient descent of sigma
            if (alpha_sigma != 0):
                D_sigma = np.zeros((n, n), float)
                for j in range(0, n):
                    der_sigma = np.zeros((n, n), float)# partial derivative of Sigma w.r.t. sigma
                    for p in range(0, n):
                        for q in range(0, n):
                            if p==j and q==j:
                                der_sigma[p][q] = 2 * sigma[j]
                            elif p==j:
                                der_sigma[p][q] = rho[p][q] * sigma[q]
                            elif q==j:
                                der_sigma[p][q] = rho[p][q] * sigma[p]
                            else:
                                der_sigma[p][q] = 0
                    D_sigma = -1.0 * temp_inv_Sigma * np.matrix(der_sigma) * temp_inv_Sigma
                    for i in range(0, m):
                        second_matrix = np.zeros((l, 1), float)
        #                print second_matrix
        #                print l
                        for k in range(0, l):
                            second_matrix[k] = np.matrix(X[i, :] - L[k, :]) * np.matrix(D_sigma) * np.matrix(X[i, :] - L[k, :]).transpose()
                        grad_sigma[j] = grad_sigma[j] + (y[i] - predictions[i]) * np.multiply(np.matrix(F[i, 1::]), np.matrix(theta[1::])) * np.matrix(second_matrix)

                for k in range(0, n):
                    sigma[k] = sigma[k] - alpha_sigma * 1.0/(2*m) * grad_sigma[k]
            # ---------------------------------------------end GD of sigma --------------------------------------
            
            # Gradient descent of rho
            if (alpha_rho != 0):
                D_rho = np.zeros((n, n), float)
                for j_p in range(0, n):
                    for j_q in range(j_p+1, n):
                        der_rho = np.zeros((n, n), float)
                        for p in range(0, n):
                            for q in range(p+1, n):
                                if p==j_p and q==j_q:
                                    der_rho[p][q] = sigma[p] * sigma[q]
                                    der_rho[q][p] = der_rho[p][q]
                                else :
                                    der_rho[p][q] = 0
                                    der_rho[q][p] = der_rho[p][q]
                        D_rho = -1.0 * temp_inv_Sigma* np.matrix(der_rho) * temp_inv_Sigma
                        for i in range(0, m):
                            second_matrix = np.zeros((l, 1), float)
                            for k in range(0, l):
                                second_matrix[k] = np.matrix(X[i, :] - L[k, :]) * np.matrix(D_rho) * np.matrix(X[i, :] - L[k, :]).transpose()
                            grad_rho[j_p][j_q] = grad_rho[j_p][j_q] + (y[i] - predictions[i]) * np.multiply(np.matrix(F[i, 1::]), np.matrix(theta[1::])) * np.matrix(second_matrix)

                for p in range(0, n):
                    for q in range(p+1, n):
                        if(p==q):
                            rho[p][q] = 1
                        else:
                            rho[p][q] = rho[p][q] - alpha_rho * (1.0/(2*m))* grad_rho[p][q]
                            rho[q][p] = rho[p][q]
            # ---------------------------------------------end GD of rho ----------------------------------------

            # Gradient descent of theta
            grad_theta = np.zeros((l+1, 1), float)
            for k in range(0, l+1):
                for i in range(0, m):
                    grad_theta[k] = grad_theta[k] + (-1.0/m)* (y[i] - predictions[i]) * F[i][k]
            
            for k in range(0, l+1):
                theta[k] = theta[k] - alpha_theta * grad_theta[k]
            
            # ---------------------------------------------end GD of theta -----------

            if inc==0:
                finish = time.time()
    
    return theta, sigma, rho, J
